CoreDNS continues to look for zone data files in old root after root directive added

I initially configured a test zone like so:

```
. {
        proxy . 8.8.8.8
        cache
        chaos CoreDNS-1.0.5 cricket@infoblox.com
        log
        errors

        file db.foo.example foo.example {
                transfer to *
        }
}
```

That worked fine.  But then I moved db.foo.example into /etc/coredns, added a new RR to the zone, added the statement "root /etc/coredns" to Corefile, and then sent SIGUSR1 to coredns.  coredns read the modified zone data file (yay!) but continues to look in the old location, according to coredns's error output:

```
2018/02/08 18:30:45 [INFO] Reloading
2018/02/08 18:30:45 [INFO] Reloading complete
2018/02/08 18:30:56 [INFO] Successfully reloaded zone "foo.example." in "/etc/coredns/db.foo.example" with serial 2018020801
127.0.0.1 - [08/Feb/2018:18:30:56 +0000] 35360 "A IN www.foo.example. udp 57 false 4096" NOERROR qr,aa,rd,ra 93 224.587µs
2018/02/08 18:31:14 [INFO] Successfully reloaded zone "foo.example." in "/etc/coredns/db.foo.example" with serial 2018020801
2018/02/08 18:31:26 [ERROR] Failed to open zone "foo.example." in "db.foo.example": open db.foo.example: no such file or directory
2018/02/08 18:31:45 [ERROR] Failed to open zone "foo.example." in "db.foo.example": open db.foo.example: no such file or directory
```

(Note the two errors at the end.)

Perhaps coredns isn't deleting the old configuration for the zone?

--- a/plugin/auto/zone.go
+++ b/plugin/auto/zone.go
@@ -60,8 +60,8 @@ func (z *Zones) Add(zo *file.Zone, name string) {
 func (z *Zones) Remove(name string) {
 	z.Lock()
 
-	if zo, ok := z.Z[name]; ok && !zo.NoReload {
-		zo.ReloadShutdown <- true
+	if zo, ok := z.Z[name]; ok {
+		zo.OnShutdown()
 	}
 
 	delete(z.Z, name)

--- a/plugin/file/reload.go
+++ b/plugin/file/reload.go
@@ -47,7 +47,7 @@ func (z *Zone) Reload() error {
 				log.Printf("[INFO] Successfully reloaded zone %q in %q with serial %d", z.origin, z.file, z.Apex.SOA.Serial)
 				z.Notify()
 
-			case <-z.ReloadShutdown:
+			case <-z.reloadShutdown:
 				tick.Stop()
 				return
 			}

--- a/plugin/file/setup.go
+++ b/plugin/file/setup.go
@@ -40,7 +40,7 @@ func setup(c *caddy.Controller) error {
 	}
 	for _, n := range zones.Names {
 		z := zones.Z[n]
-		z.OnShutdown()
+		c.OnShutdown(z.OnShutdown)
 	}
 
 	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {

--- a/plugin/file/setup.go
+++ b/plugin/file/setup.go
@@ -38,6 +38,10 @@ func setup(c *caddy.Controller) error {
 			return nil
 		})
 	}
+	for _, n := range zones.Names {
+		z := zones.Z[n]
+		z.OnShutdown()
+	}
 
 	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
 		return File{Next: next, Zones: zones}

--- a/plugin/file/setup_test.go
+++ b/plugin/file/setup_test.go
@@ -53,6 +53,20 @@ func TestFileParse(t *testing.T) {
 			false,
 			Zones{Names: []string{"10.in-addr.arpa."}},
 		},
+		{
+			`file ` + zoneFileName1 + ` example.net. {
+				upstream a
+			}`,
+			true,
+			Zones{Names: []string{}},
+		},
+		{
+			`file ` + zoneFileName1 + ` example.net. {
+				no_rebloat
+			}`,
+			true,
+			Zones{Names: []string{}},
+		},
 	}
 
 	for i, test := range tests {

--- a/plugin/file/shutdown.go
+++ b/plugin/file/shutdown.go
@@ -1,0 +1,9 @@
+package file
+
+// OnShutdown shuts down any running go-routines for this zone.
+func (z *Zone) OnShutdown() error {
+	if !z.NoReload {
+		z.reloadShutdown <- true
+	}
+	return nil
+}

--- a/plugin/file/zone.go
+++ b/plugin/file/zone.go
@@ -29,7 +29,7 @@ type Zone struct {
 
 	NoReload       bool
 	reloadMu       sync.RWMutex
-	ReloadShutdown chan bool
+	reloadShutdown chan bool
 	Upstream       upstream.Upstream // Upstream for looking up names during the resolution process
 }
 
@@ -49,7 +49,7 @@ func NewZone(name, file string) *Zone {
 		file:           path.Clean(file),
 		Tree:           &tree.Tree{},
 		Expired:        new(bool),
-		ReloadShutdown: make(chan bool),
+		reloadShutdown: make(chan bool),
 	}
 	*z.Expired = false
 

